aten/src/ATen/ParallelNative.cpp

```diff
@@ -140,21 +141,28 @@ void _parallel_run(
         ParallelRegionGuard guard(task_id);
         f(local_start, local_end, task_id);
       } catch (...) {
-        if (!err_flag.test_and_set()) {
-          eptr = std::current_exception();
+        if (!state.err_flag.test_and_set()) {
+          state.eptr = std::current_exception();
         }
       }
     }
-    futures[task_id]->markCompleted();
+
+    if (state.remaining.fetch_sub(1) == 1)
```

Comment:
Wondering: would it be ok to use the standard condvar idiom here, where the mutex goes in the state struct, and then acquire/release mutex for the decrement?

Maybe I'm wrong here (too much WFH), but consider the following sequence:

```
waiter:    std::unique_lock<std::mutex> lk(mutex);
                while(state.remaining.load()) {  // remaining is 1 
task:       if (state.remaining.fetch_sub(1) == 1)
                  state.cv.notify_one();  // notifies, but no waiter list.
waiter:    state.cv.wait(lk);  // drops lock, waits forever
```

Acquiring the mutex seems to help here by blocking the decrement from happening between the critical section between the load() call and the cv.wait(lk) call. I think?

If we really want in the future, we can copy-paste in the folly EventCount impl somewhere in c10, which does some more internal gymnastics for correctness. But I think we already got 90% of the benefit here by switching away from futures.

Reply:
@jjlilley I was thinking of exactly the same thing: just the order of operations is not sufficient.
Updated the diff to use condvar, mutex, but regular non-atomic remaining counter, since it will only be modified with the lock held.